crates/matrix-sdk-ui/src/sync_service/mod.rs

```diff
@@ -72,118 +104,295 @@ impl SyncService {
         self.room_list_service.clone()
     }
 
-    /// Observe the current state of the application.
-    ///
-    /// See also [`SyncServiceState`].
+    /// Returns the state of the sync service.
     pub fn state(&self) -> Subscriber<SyncServiceState> {
-        self.state_observer.subscribe()
+        self.state.subscribe()
     }
 
-    /// Start (or restart) the underlying sliding syncs.
-    ///
-    /// This can be called multiple times safely:
-    /// - if the stream is still properly running, it won't be restarted.
-    /// - if the stream has been aborted before, it will be properly cleaned up
-    ///   and restarted.
-    pub async fn start(&self) -> Result<(), Error> {
-        let room_list = self.room_list_service.clone();
+    fn spawn_scheduler_task(
+        &self,
+        mut receiver: Receiver<TerminationReport>,
+    ) -> impl Future<Output = ()> {
+        let encryption_sync_task = self.encryption_sync_task.clone();
         let encryption_sync = self.encryption_sync.clone();
-        let state_observer = self.state_observer.clone();
+        let room_list_service = self.room_list_service.clone();
+        let room_list_task = self.room_list_task.clone();
+        let state = self.state.clone();
+
+        async move {
+            let Some(report) = receiver.recv().await else {
+                info!("sync service scheduler: channel has been closed?");
+                return;
+            };
+
+            let (stop_room_list, stop_encryption) = match &report.origin {
+                TerminationOrigin::EncryptionSync => (true, false),
+                TerminationOrigin::RoomList => (false, true),
+                TerminationOrigin::Scheduler => (true, true),
+            };
+
+            // Put both tasks in the same state: abort and drop them.
+            if let Some(task) = room_list_task.lock().unwrap().take() {
+                task.abort();
+            }
+            if let Some(task) = encryption_sync_task.lock().unwrap().take() {
+                task.abort();
+            }
```

Comment:
I would move this _after_ the `if stop_room_list` and `if stop_encryption` blocks.

Reply:
Out of curiosity, any reason why? I wasn't sure in which order to do the things. If we abort later, we give slightly more time (nanoseconds, really) for the services to finish their current iteration, but that shouldn't change things a lot.

Reply:
(Stopping the sliding syncs first makes it possible that they'll send other termination reports to the scheduler task, which only listens to one termination report before exiting. I am not sure this can cause issues if we did that, but aborting first sounds slightly safer because of that.)

Reply:
Resolved: we're stopping the services first and then wait for the tasks to gracefully shut down. This is especially important for the room list service which will transition to `Terminated`, so its next restart will be proper.